### ISSUE TITLE
Using _1 as an Ident Causes SPIRV -> WGSL to generate empty/invalid names

### ISSUE THREAD

**Issue by cwfitzgerald** [gfx-rs/naga#1499]
The following declaration:

```glsl
layout(set = 0, binding = INDIRECT_BUFFER_BINDING, std430) restrict buffer IndirectBuffer {
    uint draw_call_count;
    uint _2;
    uint _3;
    uint _4;
    IndirectCall indirect_call[];
};
```

when compiled to spirv with glslc, and transpiled to wgsl using latest naga, results in some really weird (and invalid) names.

```
[[block]]
struct IndirectBuffer {
    draw_call_count: u32;
    : u32;
    1: u32;
    2: u32;
    indirect_call: [[stride(20)]] array<IndirectCall>;
};
```

Below is source GLSL (compiled with `-DPREFIX_OUTPUT`), spirv, and wgsl.

[naga-bug.zip](https://github.com/gfx-rs/naga/files/7448001/naga-bug.zip)

I haven't assigned a lang/area for this as I have no idea whose "fault" it is.

**Comment by kvark** [comment]
Somewhat similar to #1483
We need to teach `Namer` that names starting with digits are bad.

**Comment by Gordon-F** [comment]
> We need to teach `Namer` that names starting with digits are bad.

We already have this in namer - `reserved_prefixes`. Should we add `_` as a reserved prefix for all backends?

**Comment by jimblandy** [comment]
This is all probably fallout from #1465, "Use Cow to reduce allocation in proc::Namer", merged as ed4cc8e4. That change avoided introducing unnecessary "_" suffixes, but apparently goes a little overboard, and our tests don't cover it.

**Comment by Gordon-F** [comment]
> We already have this in namer - reserved_prefixes. Should we add _ as a reserved prefix for all backends?

@jimblandy Are you still working on this? I think this can be fixed pretty easy, like I described in the quote (correct me if I am wrong). I can send a PR.

**Comment by kvark** [comment]
we shouldn't add "_" as reserved for all backends. In fact, it's no longer reserved even for WGSL

**Comment by jimblandy** [comment]
This was fixed by 8bc3aa82. Sorry I didn't close this.

**Comment by jimblandy** [comment]
Oh, this transmuted into a new problem: we're running into the reserved prefix.

**Comment by jimblandy** [comment]
@Gordon-F Sorry for not checking back on this to make sure it was really fixed. Yes, a PR to implement the new WGSL rules around underscores in identifiers would be welcome.

**Comment by Gordon-F** [comment]
Fixed by #1547.
